rust/protocol/src/lib.rs

```diff
@@ -3,28 +3,41 @@
 // SPDX-License-Identifier: AGPL-3.0-only
 //
 
+//! Rust implementation of the **[Signal Protocol]** for asynchronous
+//! forward-secret public-key cryptography.
+//!
+//! In particular, this library implements operations conforming to the following specifications:
+//! - the **[X3DH]** key agreement protocol,
+//! - the **[Double Ratchet]** *(Axolotl)* messaging protocol,
+//! - the **[Sesame]** session agreement protocol.
+//!
+//! [Signal Protocol]: https://signal.org/
+//! [X3DH]: https://signal.org/docs/specifications/x3dh/
+//! [Double Ratchet]: https://signal.org/docs/specifications/doubleratchet/
+//! [Sesame]: https://signal.org/docs/specifications/sesame/
+
 #![warn(clippy::unwrap_used)]
 #![deny(unsafe_code)]
 
-mod address;
-mod consts;
-mod crypto;
-mod curve;
+pub mod address;
+pub mod consts;
+pub mod crypto;
+pub mod curve;
 pub mod error;
-mod fingerprint;
-mod group_cipher;
-mod identity_key;
-mod kdf;
-mod proto;
-mod protocol;
-mod ratchet;
-mod sealed_sender;
-mod sender_keys;
-mod session;
-mod session_cipher;
-mod state;
-mod storage;
-mod utils;
+pub mod fingerprint;
+pub mod group_cipher;
+pub mod identity_key;
+pub mod kdf;
+pub mod proto;
+pub mod protocol;
+pub mod ratchet;
+pub mod sealed_sender;
+pub mod sender_keys;
+pub mod session;
+pub mod session_cipher;
+pub mod state;
+pub mod storage;
+pub mod utils;
```

Comment:
I know you mentioned that this is one of your goals in #285, but even if we do decide to publicize the module structure of the library, it won't be by making every single submodule public. (And it might not be with these names either.) We should think about the organization that makes sense for the types that really do make sense to export (and possibly talk about it too, either in #285 or on https://community.signalusers.org).

Reply:
That's a great amendment to the goals I wrote in #285 and I will link this comment there and follow up in https://community.signalusers.org/! I will avoid changing any of the module structure right now except to make the `address` module `pub` -- does that sound reasonable?

Reply:
Sorry, but I'd really rather keep the various kinds of changes separate. Right now we have a flat structure for our public API, so *any* changes indicate a new design. Therefore rather than say "okay, we definitely want this module but I'm not sure about the others", I'd rather look at some possible organizations all together (flat, today's groupings, proposed future groupings) and pick one. (And similarly I think it would make sense to audit for `pub(crate)` first, at least on types, so that we know what the set of actually-public types is.)

Reply:
Ok, that's very easy to follow! I think that `cargo doc` actually has flags to build things regardless of `pub(crate)` status, which I'm realizing is the only motivation to mess with the module structure at all. Thank you for the prompt and complete response!